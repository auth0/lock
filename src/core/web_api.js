import auth0 from 'auth0-js';
import Auth0LegacyAPIClient from './web_api/legacy_api'
import Auth0APIClient from './web_api/p2_api'

class Auth0WebAPI {
  constructor() {
    this.clients = {};
  }

  setupClient(lockID, clientID, domain, opts) {

    const hostedLoginPage = window.location.host === domain;
    // when it is used on on the hosted login page, it shouldn't use popup mode
    opts.redirect = hostedLoginPage ? true : opts.redirect;

    opts.oidcConformant = opts.oidcConformant || false;

    // for cordova and electron we should force popup without SSO so it uses
    // /ro or /oauth/token for DB connections
    if (window && (!!window.cordova || !!window.electron)) {
      opts.redirect = false;
      opts.sso = false;
    }

    // when it is used on on the hosted login page, it should use the legacy mode
    // (usernamepassword/login) in order to continue the transaction after authentication
    if (hostedLoginPage || !opts.oidcConformant) {
      this.clients[lockID] = new Auth0LegacyAPIClient(clientID, domain, opts);
    } else {
      this.clients[lockID] = new Auth0APIClient(lockID, clientID, domain, opts);
    }
  }

  logIn(lockID, options, authParams, cb) {
    this.clients[lockID].logIn(options, authParams, cb);
  }

  signOut(lockID, query) {
    this.clients[lockID].logout(query);
  }

  signUp(lockID, options, cb) {
    this.clients[lockID].signUp(options, cb);
  }

  resetPassword(lockID, options, cb) {
    this.clients[lockID].resetPassword(options, cb);
  }

  startPasswordless(lockID, options, cb) {
    this.clients[lockID].startPasswordless(options, cb);
  }

  parseHash(lockID, hash = '', cb) {
    return this.clients[lockID].parseHash(hash, cb);
  }

  getUserInfo(lockID, token, callback) {
    return this.clients[lockID].getUserInfo(token, callback);
  }

  getProfile(lockID, token, callback) {
    return this.clients[lockID].getProfile(token, callback);
  }

  getSSOData(lockID, ...args) {
    return this.clients[lockID].getSSOData(...args);
  }

  getUserCountry(lockID, cb) {
    return this.clients[lockID].getUserCountry((err, data) => cb(err, data && data.countryCode));
  }
}

export default new Auth0WebAPI();
