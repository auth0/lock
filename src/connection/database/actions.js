import Immutable, { Map } from 'immutable';
import { getEntity, read, swap, updateEntity } from '../../store/index';
import webApi from '../../core/web_api';
import { closeLock, logIn as coreLogIn, logInSuccess, validateAndSubmit } from '../../core/actions';
import * as l from '../../core/index';
import * as c from '../../field/index';
import {
  databaseConnectionName,
  databaseConnectionRequiresUsername,
  databaseLogInWithEmail,
  hasScreen,
  setScreen,
  shouldAutoLogin,
  toggleTermsAcceptance as switchTermsAcceptance,
  additionalSignUpFields
} from './index';
import * as i18n from '../../i18n';

export function logIn(id, needsMFA = false) {
  const m = read(getEntity, 'lock', id);
  const usernameField = databaseLogInWithEmail(m) ? 'email' : 'username';
  const username = c.getFieldValue(m, usernameField);
  const params = {
    connection: databaseConnectionName(m),
    username: username,
    password: c.getFieldValue(m, 'password')
  };

  const fields = [usernameField, 'password'];

  const mfaCode = c.getFieldValue(m, 'mfa_code');
  if (needsMFA) {
    params['mfa_code'] = mfaCode;
    fields.push('mfa_code');
  }

  coreLogIn(id, fields, params, (id, error, fields, next) => {
    if (error.error === 'a0.mfa_required') {
      return showLoginMFAActivity(id);
    }

    return next();
  });
}

export function signUp(id) {
  const m = read(getEntity, 'lock', id);
  const fields = ['email', 'password'];
  if (databaseConnectionRequiresUsername(m)) fields.push('username');
  additionalSignUpFields(m).forEach(x => fields.push(x.get('name')));

  validateAndSubmit(id, fields, m => {
    const params = {
      connection: databaseConnectionName(m),
      email: c.getFieldValue(m, 'email'),
      password: c.getFieldValue(m, 'password'),
      autoLogin: shouldAutoLogin(m)
    };

    if (databaseConnectionRequiresUsername(m)) {
      params.username = c.getFieldValue(m, 'username');
    }

    if (!additionalSignUpFields(m).isEmpty()) {
      params.user_metadata = {};
      additionalSignUpFields(m).forEach(x => {
        params.user_metadata[x.get('name')] = c.getFieldValue(m, x.get('name'));
      });
    }

    webApi.signUp(id, params, (error, result, popupHandler, ...args) => {
      if (error) {
        if (!!popupHandler) {
          popupHandler._current_popup.kill();
        }
        setTimeout(() => signUpError(id, error), 250);
      } else {
        signUpSuccess(id, result, popupHandler, ...args);
      }
    });
  });
}

function signUpSuccess(id, result, popupHandler) {
  const lock = read(getEntity, 'lock', id);

  if (shouldAutoLogin(lock)) {
    swap(updateEntity, 'lock', id, m => m.set('signedUp', true));

    // TODO: check options, redirect is missing
    const options = {
      connection: databaseConnectionName(lock),
      username: c.email(lock),
      password: c.password(lock)
    };

    if (!!popupHandler) {
      options.popupHandler = popupHandler;
    }

    return webApi.logIn(id, options, l.auth.params(lock).toJS(), (error, ...args) => {
      if (error) {
        setTimeout(() => autoLogInError(id, error), 250);
      } else {
        logInSuccess(id, ...args);
      }
    });
  }

  const autoclose = l.ui.autoclose(lock);

  if (!autoclose) {
    swap(updateEntity, 'lock', id, lock => l.setSubmitting(lock, false).set('signedUp', true));
  } else {
    closeLock(id, false);
  }
}

function signUpError(id, error) {
  const m = read(getEntity, 'lock', id);

  const invalidPasswordKeys = {
    PasswordDictionaryError: 'password_dictionary_error',
    PasswordNoUserInfoError: 'password_no_user_info_error',
    PasswordStrengthError: 'password_strength_error'
  };

  const errorKey =
    (error.code === 'invalid_password' && invalidPasswordKeys[error.name]) || error.code;

  const errorMessage =
    i18n.html(m, ['error', 'signUp', errorKey]) ||
    i18n.html(m, ['error', 'signUp', 'lock.fallback']);

  l.emitEvent(m, 'signup error', error);

  swap(updateEntity, 'lock', id, l.setSubmitting, false, errorMessage);
}

function autoLogInError(id, error) {
  swap(updateEntity, 'lock', id, m => {
    const errorMessage = l.loginErrorMessage(m, error);
    if (hasScreen(m, 'login')) {
      return l.setSubmitting(setScreen(m, 'login'), false, errorMessage);
    } else {
      return l.setSubmitting(m, false, errorMessage);
    }
  });
}

export function resetPassword(id) {
  validateAndSubmit(id, ['email'], m => {
    const params = {
      connection: databaseConnectionName(m),
      email: c.getFieldValue(m, 'email')
    };

    webApi.resetPassword(id, params, (error, ...args) => {
      if (error) {
        setTimeout(() => resetPasswordError(id, error), 250);
      } else {
        resetPasswordSuccess(id);
      }
    });
  });
}

function resetPasswordSuccess(id) {
  const m = read(getEntity, 'lock', id);
  if (hasScreen(m, 'login')) {
    swap(
      updateEntity,
      'lock',
      id,
      m => setScreen(l.setSubmitting(m, false), 'login', ['']) // array with one empty string tells the function to not clear any field
    );

    // TODO: should be handled by box
    setTimeout(() => {
      const successMessage = i18n.html(m, ['success', 'forgotPassword']);
      swap(updateEntity, 'lock', id, l.setGlobalSuccess, successMessage);
    }, 500);
  } else {
    if (l.ui.autoclose(m)) {
      closeLock(id);
    } else {
      swap(updateEntity, 'lock', id, m => l.setSubmitting(m, false).set('passwordResetted', true));
    }
  }
}

function resetPasswordError(id, error) {
  const m = read(getEntity, 'lock', id);

  const errorMessage =
    i18n.html(m, ['error', 'forgotPassword', error.code]) ||
    i18n.html(m, ['error', 'forgotPassword', 'lock.fallback']);

  swap(updateEntity, 'lock', id, l.setSubmitting, false, errorMessage);
}

export function showLoginActivity(id, fields = ['password']) {
  swap(updateEntity, 'lock', id, setScreen, 'login', fields);
}

export function showSignUpActivity(id, fields = ['password']) {
  swap(updateEntity, 'lock', id, setScreen, 'signUp', fields);
}

export function showResetPasswordActivity(id, fields = ['password']) {
  swap(updateEntity, 'lock', id, setScreen, 'forgotPassword', fields);
}

export function cancelResetPassword(id) {
  return showLoginActivity(id);
}

export function cancelMFALogin(id) {
  return showLoginActivity(id);
}

export function toggleTermsAcceptance(id) {
  swap(updateEntity, 'lock', id, switchTermsAcceptance);
}

export function showLoginMFAActivity(id, fields = ['mfa_code']) {
  swap(updateEntity, 'lock', id, setScreen, 'mfaLogin', fields);
}
