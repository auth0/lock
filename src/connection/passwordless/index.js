import Immutable, { List, Map } from 'immutable';
import * as l from '../../core/index';
import { clearFields } from '../../field/index';
import { initLocation } from '../../field/phone_number';
import { dataFns } from '../../utils/data_utils';
const { get, initNS, tget, tremove, tset } = dataFns(['passwordless']);
import webAPI from '../../core/web_api';
import sync from '../../sync';

export function initPasswordless(m, opts) {
  // TODO: validate opts

  const send = typeof opts.sendCode === 'boolean' && opts.sendCode ? 'code' : 'link';

  m = initNS(m, Map({ send: send }));
  if (opts.defaultLocation && typeof opts.defaultLocation === 'string') {
    m = initLocation(m, opts.defaultLocation.toUpperCase());
  } else {
    m = sync(m, 'location', {
      recoverResult: 'US',
      syncFn: (m, cb) => webAPI.getUserCountry(l.id(m), cb),
      successFn: (m, result) => initLocation(m, result)
    });
  }

  return m;
}

function setResendStatus(m, value) {
  // TODO: check value
  return tset(m, 'resendStatus', value);
}

export function setResendSuccess(m) {
  return setResendStatus(m, 'success');
}

export function resendSuccess(m) {
  return resendStatus(m) == 'success';
}

export function setResendFailed(m) {
  return setResendStatus(m, 'failed');
}

export function resendFailed(m) {
  return resendStatus(m) == 'failed';
}

export function resendOngoing(m) {
  return resendStatus(m) == 'ongoing';
}

export function resend(m) {
  if (resendAvailable(m)) {
    return setResendStatus(m, 'ongoing');
  } else {
    return m;
  }
}

function resendStatus(m) {
  return tget(m, 'resendStatus', 'waiting');
}

export function resendAvailable(m) {
  return resendStatus(m) == 'waiting' || resendStatus(m) == 'failed';
}

export function restartPasswordless(m) {
  // TODO: maybe we can take advantage of the transient fields
  m = tremove(m, 'passwordlessStarted');
  m = tremove(m, 'resendStatus'); // only for link
  m = clearFields(m, ['vcode']); // only for code

  return l.clearGlobalError(m);
}

export function send(m) {
  return get(m, 'send', isEmail(m) ? 'link' : 'code');
}

export function isSendLink(m) {
  return send(m) === 'link';
}

export function setPasswordlessStarted(m, value) {
  return tset(m, 'passwordlessStarted', value);
}

export function passwordlessStarted(m) {
  return tget(m, 'passwordlessStarted', false);
}

export function passwordlessConnection(m) {
  return (
    l.connections(m, 'passwordless', 'email').get(0) ||
    l.connections(m, 'passwordless', 'sms').get(0) ||
    new Map()
  );
}

export function isEmail(m) {
  const c = passwordlessConnection(m);
  return c.isEmpty() ? undefined : c.get('strategy') === 'email';
}
