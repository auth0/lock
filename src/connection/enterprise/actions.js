import { getEntity, read, swap, updateEntity } from '../../store/index';
import {
  enterpriseActiveFlowConnection,
  isHRDActive,
  matchConnection,
  toggleHRD
} from '../enterprise';
import { getFieldValue, hideInvalidFields } from '../../field/index';
import { logIn as coreLogIn } from '../../core/actions';

// TODO: enterprise connections should not depend on database
// connections. However, we now allow a username input to contain also
// an email and this information is in the database module. We should
// make this information flow from the UI (like we do for the startHRD
// function). Including this dependency here allows us to do that
// incrementally.
import { databaseLogInWithEmail } from '../database/index';

export function startHRD(id, email) {
  swap(updateEntity, "lock", id, toggleHRD, email);
}

export function cancelHRD(id) {
  swap(updateEntity, "lock", id, m => {
    m = toggleHRD(m, false);
    m = hideInvalidFields(m);
    return m;
  });
}

export function logIn(id) {
  const m = read(getEntity, "lock", id);
  const email = getFieldValue(
    m,
    databaseLogInWithEmail(m) ? "email" : "username"
  );
  const ssoConnection = matchConnection(m, email);

  if (ssoConnection && !isHRDActive(m)) {
    return logInSSO(id, ssoConnection);
  }

  logInActiveFlow(id);
}

function logInActiveFlow(id) {
  const m = read(getEntity, "lock", id);
  const usernameField = isHRDActive(m) || !databaseLogInWithEmail(m)
    ? "username"
    : "email";

  const username = getFieldValue(m, usernameField);
  const connection = enterpriseActiveFlowConnection(m); 

  coreLogIn(id, ["password", usernameField], {
    connection: connection ? connection.get("name") : null,
    username: username,
    password: getFieldValue(m, "password"),
    login_hint: username
  });
}

function logInSSO(id, connection) {
  const m = read(getEntity, "lock", id);
  const field = databaseLogInWithEmail(m) ? "email" : "username";
  coreLogIn(id, [field], {
    connection: connection.get("name"),
    login_hint: getFieldValue(m, field)
  });
}
