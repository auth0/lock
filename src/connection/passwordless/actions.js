import { Map } from 'immutable';
import { read, getEntity, swap, updateEntity } from '../../store/index';
import { closeLock, logIn as coreLogIn, validateAndSubmit, logInSuccess } from '../../core/actions';
import webApi from '../../core/web_api';
import * as c from '../../field/index';
import * as l from '../../core/index';
import {
  isEmail,
  isSendLink,
  resend,
  restartPasswordless,
  send,
  setPasswordlessStarted,
  setResendFailed,
  setResendSuccess
} from './index';
import { phoneNumberWithDiallingCode } from '../../field/phone_number';
import * as i18n from '../../i18n';

function getErrorMessage(m, error) {
  let key = error.error;

  if (
    error.error === 'sms_provider_error' &&
    (error.description || '').indexOf('(Code: 21211)') > -1
  ) {
    key = 'bad.phone_number';
  }

  return (
    i18n.html(m, ['error', 'passwordless', key]) ||
    i18n.html(m, ['error', 'passwordless', 'lock.fallback'])
  );
}

export function requestPasswordlessEmail(id) {
  validateAndSubmit(id, ['email'], m => {
    sendEmail(m, requestPasswordlessEmailSuccess, requestPasswordlessEmailError);
  });
}

export function requestPasswordlessEmailSuccess(id) {
  swap(updateEntity, 'lock', id, m => {
    m = l.setSubmitting(m, false);
    return setPasswordlessStarted(m, true);
  });
}

export function requestPasswordlessEmailError(id, error) {
  const m = read(getEntity, 'lock', id);
  const errorMessage = getErrorMessage(m, error);
  return swap(updateEntity, 'lock', id, l.setSubmitting, false, errorMessage);
}

export function resendEmail(id) {
  swap(updateEntity, 'lock', id, resend);
  const m = read(getEntity, 'lock', id);
  sendEmail(m, resendEmailSuccess, resendEmailError);
}

function resendEmailSuccess(id) {
  swap(updateEntity, 'lock', id, setResendSuccess);
}

function resendEmailError(id, error) {
  swap(updateEntity, 'lock', id, setResendFailed);
}

function sendEmail(m, successFn, errorFn) {
  const params = {
    connection: 'email',
    email: c.getFieldValue(m, 'email'),
    send: send(m)
  };

  if (isSendLink(m) && !l.auth.params(m).isEmpty()) {
    params.authParams = l.auth.params(m).toJS();
  }

  webApi.startPasswordless(l.id(m), params, error => {
    if (error) {
      setTimeout(() => errorFn(l.id(m), error), 250);
    } else {
      successFn(l.id(m));
    }
  });
}

export function sendSMS(id) {
  validateAndSubmit(id, ['phoneNumber'], m => {
    const params = {
      connection: 'sms',
      phoneNumber: phoneNumberWithDiallingCode(m),
      send: send(m)
    };
    webApi.startPasswordless(id, params, error => {
      if (error) {
        setTimeout(() => sendSMSError(id, error), 250);
      } else {
        sendSMSSuccess(id);
      }
    });
  });
}

function sendSMSSuccess(id) {
  swap(updateEntity, 'lock', id, m => {
    m = l.setSubmitting(m, false);
    m = setPasswordlessStarted(m, true);
    return m;
  });
}

function sendSMSError(id, error) {
  const m = read(getEntity, 'lock', id);
  const errorMessage = getErrorMessage(m, error);
  return swap(updateEntity, 'lock', id, l.setSubmitting, false, errorMessage);
}

export function logIn(id) {
  const m = read(getEntity, 'lock', id);
  const authParams = l.auth.params(m).toJS();
  const params = {
    verificationCode: c.getFieldValue(m, 'vcode'),
    ...authParams
  };
  if (isEmail(m)) {
    params.connection = 'email';
    params.email = c.getFieldValue(m, 'email');
  } else {
    params.connection = 'sms';
    params.phoneNumber = phoneNumberWithDiallingCode(m);
  }
  swap(updateEntity, 'lock', id, l.setSubmitting, true);
  webApi.passwordlessVerify(id, params, (error, result) => {
    let errorMessage;
    if (error) {
      const m = read(getEntity, 'lock', id);
      errorMessage = getErrorMessage(m, error);
      if (error.logToConsole) {
        console.error(error.description);
      }
      return swap(updateEntity, 'lock', id, l.setSubmitting, false, errorMessage);
    } else {
      return logInSuccess(id, result);
    }
  });
}

export function restart(id) {
  swap(updateEntity, 'lock', id, restartPasswordless);
}
