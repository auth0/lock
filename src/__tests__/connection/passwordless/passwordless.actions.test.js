import passwordless from 'connection/passwordless/actions';
import { expectMockToMatch } from 'testUtils';

jest.useFakeTimers();

describe('passwordless actions', () => {
  let mockFns;
  let actions;
  beforeEach(() => {
    jest.resetModules();

    jest.mock('connection/passwordless/index', () => ({
      isEmail: jest.fn(),
      isSendLink: jest.fn(),
      resend: 'resend',
      restartPasswordless: jest.fn(),
      send: () => 'send',
      setPasswordlessStarted: jest.fn(),
      setResendFailed: jest.fn(),
      setResendSuccess: jest.fn()
    }));
    jest.mock('field/phone_number', () => ({
      phoneNumberWithDiallingCode: () => 'phoneNumberWithDiallingCode'
    }));
    jest.mock('field/index', () => ({
      getFieldValue: (m, field) => field
    }));
    jest.mock('core/web_api', () => ({
      startPasswordless: jest.fn(),
      passwordlessVerify: jest.fn()
    }));
    jest.mock('core/actions', () => ({
      closeLock: jest.fn(),
      logIn: jest.fn(),
      validateAndSubmit: jest.fn(),
      logInSuccess: jest.fn()
    }));
    jest.mock('i18n', () => ({ html: (_, keys) => keys.join(',') }));
    jest.mock('core/index', () => ({
      id: () => 'id',
      setSubmitting: jest.fn(m => m),
      auth: {
        params: () => ({
          toJS: () => ({
            auth: 'params'
          })
        })
      }
    }));
    jest.mock('store/index', () => ({
      read: jest.fn(() => 'model'),
      getEntity: 'getEntity',
      swap: jest.fn(),
      updateEntity: 'updateEntity'
    }));

    actions = require('connection/passwordless/actions');
  });
  describe('requestPasswordlessEmail()', () => {
    it('calls validateAndSubmit()', () => {
      actions.requestPasswordlessEmail('id');
      expectMockToMatch(require('core/actions').validateAndSubmit, 1);
    });
    it('calls startPasswordless', () => {
      actions.requestPasswordlessEmail('id');
      require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
      expectMockToMatch(require('core/web_api').startPasswordless, 1);
    });
    it('calls setPasswordlessStarted() on success', () => {
      actions.requestPasswordlessEmail('id');
      require('core/actions').validateAndSubmit.mock.calls[0][2]('model');

      require('core/web_api').startPasswordless.mock.calls[0][2](null);

      const { swap } = require('store/index');
      expectMockToMatch(swap, 1);

      swap.mock.calls[0][3]('model');
      expectMockToMatch(require('core/index').setSubmitting, 1);
      expectMockToMatch(require('connection/passwordless/index').setPasswordlessStarted, 1);
    });
    describe('normalizes the error message', () => {
      it('with a generic error', () => {
        actions.requestPasswordlessEmail('id');
        require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
        const error = new Error('foobar');
        error.error = 'some_error_code';
        require('core/web_api').startPasswordless.mock.calls[0][2](error);

        jest.runAllTimers();

        const { read, swap } = require('store/index');
        expectMockToMatch(read, 1);
        expectMockToMatch(swap, 1);
      });
      it('with a sms_provider_error error and description includes (Code: 21211)', () => {
        actions.requestPasswordlessEmail('id');
        require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
        const error = new Error('foobar');
        error.error = 'sms_provider_error';
        error.description = 'something (Code: 21211)';
        require('core/web_api').startPasswordless.mock.calls[0][2](error);

        jest.runAllTimers();

        const { read, swap } = require('store/index');
        expectMockToMatch(swap, 1);
      });
    });
  });
  describe('resendEmail()', () => {
    it('calls setResendSuccess() on success', () => {
      actions.resendEmail('id');

      const { read, swap } = require('store/index');
      expectMockToMatch(read, 1);
      expectMockToMatch(swap, 1);

      require('core/web_api').startPasswordless.mock.calls[0][2](null);

      swap.mock.calls[1][3]('model');
      expectMockToMatch(require('connection/passwordless/index').setResendSuccess, 1);
    });
    it('calls setResendFailed on error', () => {
      actions.resendEmail('id');

      const { read, swap } = require('store/index');
      expectMockToMatch(read, 1);
      expectMockToMatch(swap, 1);

      require('core/web_api').startPasswordless.mock.calls[0][2](new Error('foobar'));
      jest.runAllTimers();
      swap.mock.calls[1][3]('model');
      expectMockToMatch(require('connection/passwordless/index').setResendFailed, 1);
    });
  });
  describe('sendSMS()', () => {
    it('calls validateAndSubmit()', () => {
      actions.sendSMS('id');
      expectMockToMatch(require('core/actions').validateAndSubmit, 1);
    });
    it('calls startPasswordless', () => {
      actions.sendSMS('id');
      require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
      expectMockToMatch(require('core/web_api').startPasswordless, 1);
    });
    it('calls setPasswordlessStarted() on success', () => {
      actions.sendSMS('id');
      require('core/actions').validateAndSubmit.mock.calls[0][2]('model');

      require('core/web_api').startPasswordless.mock.calls[0][2](null);

      const { swap } = require('store/index');
      expectMockToMatch(swap, 1);

      swap.mock.calls[0][3]('model');
      expectMockToMatch(require('core/index').setSubmitting, 1);
      expectMockToMatch(require('connection/passwordless/index').setPasswordlessStarted, 1);
    });
    describe('normalizes the error message', () => {
      it('with a generic error', () => {
        actions.sendSMS('id');
        require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
        const error = new Error('foobar');
        error.error = 'some_error_code';
        require('core/web_api').startPasswordless.mock.calls[0][2](error);

        jest.runAllTimers();

        const { read, swap } = require('store/index');
        expectMockToMatch(read, 1);
        expectMockToMatch(swap, 1);
      });
      it('with a sms_provider_error error and description includes (Code: 21211)', () => {
        actions.sendSMS('id');
        require('core/actions').validateAndSubmit.mock.calls[0][2]('model');
        const error = new Error('foobar');
        error.error = 'sms_provider_error';
        error.description = 'something (Code: 21211)';
        require('core/web_api').startPasswordless.mock.calls[0][2](error);

        jest.runAllTimers();

        const { read, swap } = require('store/index');
        expectMockToMatch(swap, 1);
      });
    });
  });
  describe('login()', () => {
    it('sets setSubmitting to true', () => {
      actions.logIn('id');

      const { read, swap } = require('store/index');
      expectMockToMatch(read, 1);
      expectMockToMatch(swap, 1);
    });
    it('calls webApi.passwordlessVerify() with sms options', () => {
      actions.logIn('id');
      expectMockToMatch(require('core/web_api').passwordlessVerify, 1);
    });
    it('calls webApi.passwordlessVerify() with email options', () => {
      require('connection/passwordless/index').isEmail = () => true;
      actions.logIn('id');
      expectMockToMatch(require('core/web_api').passwordlessVerify, 1);
    });
    describe('on webApi.passwordlessVerify() callback', () => {
      it('formats error when there is an error ', () => {
        actions.logIn('id');

        const error = new Error('foobar');
        error.error = 'some_error_code';
        require('core/web_api').passwordlessVerify.mock.calls[0][2](error);

        const { swap } = require('store/index');
        expectMockToMatch(swap, 2);
      });
      it('calls logInSuccess on success', () => {
        actions.logIn('id');
        require('core/web_api').passwordlessVerify.mock.calls[0][2](null, { result: true });

        expectMockToMatch(require('core/actions').logInSuccess, 1);
      });
    });
  });
  it('restart calls restartPasswordless', () => {
    actions.restart('id');

    const { swap } = require('store/index');
    expectMockToMatch(swap, 1);
  });
});
