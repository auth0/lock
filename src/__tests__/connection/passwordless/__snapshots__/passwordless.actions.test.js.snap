// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passwordless actions login() calls webApi.passwordlessVerify() with email options 1`] = `
Array [
  "id",
  Object {
    "auth": "params",
    "connection": "email",
    "email": "email",
    "verificationCode": "vcode",
  },
  [Function],
]
`;

exports[`passwordless actions login() calls webApi.passwordlessVerify() with sms options 1`] = `
Array [
  "id",
  Object {
    "auth": "params",
    "connection": "sms",
    "phoneNumber": "phoneNumberWithDiallingCode",
    "verificationCode": "vcode",
  },
  [Function],
]
`;

exports[`passwordless actions login() on webApi.passwordlessVerify() callback calls logInSuccess on success 1`] = `
Array [
  "id",
  Object {
    "result": true,
  },
]
`;

exports[`passwordless actions login() on webApi.passwordlessVerify() callback formats error when there is an error  1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  true,
]
`;

exports[`passwordless actions login() on webApi.passwordlessVerify() callback formats error when there is an error  2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  false,
  "error,passwordless,some_error_code",
]
`;

exports[`passwordless actions login() sets setSubmitting to true 1`] = `
Array [
  "getEntity",
  "lock",
  "id",
]
`;

exports[`passwordless actions login() sets setSubmitting to true 2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  true,
]
`;

exports[`passwordless actions requestPasswordlessEmail() calls setPasswordlessStarted() on success 1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
]
`;

exports[`passwordless actions requestPasswordlessEmail() calls setPasswordlessStarted() on success 2`] = `
Array [
  "model",
  false,
]
`;

exports[`passwordless actions requestPasswordlessEmail() calls setPasswordlessStarted() on success 3`] = `
Array [
  "model",
  true,
]
`;

exports[`passwordless actions requestPasswordlessEmail() calls startPasswordless 1`] = `
Array [
  "id",
  Object {
    "connection": "email",
    "email": "email",
    "send": "send",
  },
  [Function],
]
`;

exports[`passwordless actions requestPasswordlessEmail() calls validateAndSubmit() 1`] = `
Array [
  "id",
  Array [
    "email",
  ],
  [Function],
]
`;

exports[`passwordless actions requestPasswordlessEmail() normalizes the error message with a generic error 1`] = `
Array [
  "getEntity",
  "lock",
  "id",
]
`;

exports[`passwordless actions requestPasswordlessEmail() normalizes the error message with a generic error 2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  false,
  "error,passwordless,some_error_code",
]
`;

exports[`passwordless actions requestPasswordlessEmail() normalizes the error message with a sms_provider_error error and description includes (Code: 21211) 1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  false,
  "error,passwordless,bad.phone_number",
]
`;

exports[`passwordless actions resendEmail() calls setResendFailed on error 1`] = `
Array [
  "getEntity",
  "lock",
  "id",
]
`;

exports[`passwordless actions resendEmail() calls setResendFailed on error 2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  "resend",
]
`;

exports[`passwordless actions resendEmail() calls setResendFailed on error 3`] = `
Array [
  "model",
]
`;

exports[`passwordless actions resendEmail() calls setResendSuccess() on success 1`] = `
Array [
  "getEntity",
  "lock",
  "id",
]
`;

exports[`passwordless actions resendEmail() calls setResendSuccess() on success 2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  "resend",
]
`;

exports[`passwordless actions resendEmail() calls setResendSuccess() on success 3`] = `
Array [
  "model",
]
`;

exports[`passwordless actions restart calls restartPasswordless 1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
]
`;

exports[`passwordless actions sendSMS() calls setPasswordlessStarted() on success 1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
]
`;

exports[`passwordless actions sendSMS() calls setPasswordlessStarted() on success 2`] = `
Array [
  "model",
  false,
]
`;

exports[`passwordless actions sendSMS() calls setPasswordlessStarted() on success 3`] = `
Array [
  "model",
  true,
]
`;

exports[`passwordless actions sendSMS() calls startPasswordless 1`] = `
Array [
  "id",
  Object {
    "connection": "sms",
    "phoneNumber": "phoneNumberWithDiallingCode",
    "send": "send",
  },
  [Function],
]
`;

exports[`passwordless actions sendSMS() calls validateAndSubmit() 1`] = `
Array [
  "id",
  Array [
    "phoneNumber",
  ],
  [Function],
]
`;

exports[`passwordless actions sendSMS() normalizes the error message with a generic error 1`] = `
Array [
  "getEntity",
  "lock",
  "id",
]
`;

exports[`passwordless actions sendSMS() normalizes the error message with a generic error 2`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  false,
  "error,passwordless,some_error_code",
]
`;

exports[`passwordless actions sendSMS() normalizes the error message with a sms_provider_error error and description includes (Code: 21211) 1`] = `
Array [
  "updateEntity",
  "lock",
  "id",
  [Function],
  false,
  "error,passwordless,bad.phone_number",
]
`;
